### ISSUE TITLE
5.3 data-bs-theme dark IFRAME white background

### ISSUE THREAD

**Issue by cybercussion** [twbs/bootstrap#38750]
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed issues
- [X] I have [validated](https://html5.validator.nu/) any HTML to avoid common problems
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Describe the issue

Paypal Integration for example in dark has a white background.  I attempted to find what is setting it to white and so far haven't been able to debug exactly whats doing it.  In light mode the default is transparent like expected. I've attempted to set data-bs-theme light at a level above the IFRAME with no effect.

### Reduced test cases

- use data-bs-theme="dark" on html tag
- embed PayPal integration via https://developer.paypal.com/sdk/js/reference/ or https://developer.paypal.com/demo/checkout/#/pattern/style

### What operating system(s) are you seeing the problem on?

macOS

### What browser(s) are you seeing the problem on?

Chrome, Safari, Firefox

### What version of Bootstrap are you using?

5.3.0

**Comment by cybercussion** [comment]
![image](https://github.com/twbs/bootstrap/assets/94954/dd4458a9-82b7-411b-99c5-50a19dd30146)

vs

![image](https://github.com/twbs/bootstrap/assets/94954/776cfb3b-4bdc-4b70-8bf3-b1485e23b952)

**Comment by cybercussion** [comment]
Notice it also responds similarly with

`<meta name="color-scheme" content="dark">`

**Comment by CyrilKrylatov** [comment]
My guess is since PayPal is using a `iframe` element, that's like a whole website you're loading inside yours and you won't be able to change its style from your own website. 

In other words, you can't do something like:

```css
iframe.paypal-iframe > html > body { background: black }
```

**Comment by TravsVoid** [comment]
I had the same issue but with my Disqus comment section. When I used data-bs-theme="dark" it caused the Disqus iframe to turn white instead of transparent. The only way I found to fix it was to apply the code below to the iframe which makes it transparent again for dark mode.
```
iframe {
  color-scheme: light;
}
```

**Comment by dougwaldron** [comment]
> > Anyway I don't think that's an issue from Bootstrap.
> 
> I agree with @CyrilKrylatov. From what I understand, I don't think that's an issue from Bootstrap since the `<iframe>` is independent of Bootstrap's dark mode defined in the "main" content. So closing it.

Yes and no. It's true that styles defined in Bootstrap aren't applied inside the `iframe`, but the `color-scheme` property, if set, does get passed to the `iframe`, which affects how it's displayed. See [this CodePen](https://codepen.io/dougwaldron/pen/GRzWKNG) for examples.

It seems to me that Bootstrap should set `iframe { color-scheme: normal; }` but that would be a breaking change I guess.

**Comment by lealceldeiro** [comment]
> I had the same issue but with my Disqus comment section. When I used data-bs-theme="dark" it caused the Disqus iframe to turn white instead of transparent. The only way I found to fix it was to apply the code below to the iframe which makes it transparent again for dark mode.
> 
> ```
> iframe {
>   color-scheme: light;
> }
> ```

```
[data-bs-theme="dark"] {
    color-scheme: unset;
}
```
Also does the trick

**Comment by ishields** [comment]
Ran into this same issue with our TrustPilot widget. Not sure what the impacts are setting this everywhere, so for now I went with a super scoped approach that just targets that specific iframe.

@julien-deramond This does feels like something Bootstrap should address. The issue doesn't happen when you remove bootstrap css (or if you remove dark color scheme) and is affecting a variety of different iframes (for me it was TrustPilot but there are more examples above). Think this should be reopened and considered again. Though there's a hack of sorts by overriding Boostrap style, Bootstrap shouldn't aversely affect iframes like it currently seems to be doing. 


```
.trustpilot-widget iframe {
  color-scheme: light;
}
```

**Comment by julien-deramond** [comment]
Thanks all for your feedback.
I reopen it to remind me to check all these use cases.

**Comment by gigabytes** [comment]
I was already going crazy with this; my problem was with the CometChat widget. It finally worked with:

iframe {
	color-scheme: light;
}

A huge thanks to whoever figured this out!

![Captura de tela 2024-12-16 115632](https://github.com/user-attachments/assets/9aea6c2c-4041-4544-b75a-61630086486e)

**Comment by mdo** [comment]
The white iframe background comes from the embedded third-party content's own `color-scheme`/background — Bootstrap can't style cross-origin iframe internals. Closing as out of scope — please open a new issue if you find a Bootstrap-side cause in v6-dev.

**Comment by androidacy-user** [comment]
> The white iframe background comes from the embedded third-party content's own `color-scheme`/background — Bootstrap can't style cross-origin iframe internals. Closing as out of scope — please open a new issue if you find a Bootstrap-side cause in v6-dev.

Nope. It comes from browser quirks around the css `color-scheme` set on the host document.  Iframes default to inheriting the host document here

**Comment by dougwaldron** [comment]
> The white iframe background comes from the embedded third-party content's own `color-scheme`/background — Bootstrap can't style cross-origin iframe internals.

@mdo This issue contains multiple examples showing that the `color-scheme` setting on the parent document does indeed affect how the `iframe` is displayed.

**Comment by mdo** [comment]
Sorry, I've been quickly triaging in bulk.

Okay so this is a weird one. Did some research and here's the CSS working group explanation for the issue:

> If the color scheme of an iframe differs from embedding document, iframe gets an opaque canvas background appropriate to its color scheme.
> 
> https://fvsch.com/transparent-iframes

So any mismatch results in a browser-set background.

I think setting the following should suffice, and be relatively bulletproof for us without causing more downstream issues, being placed inside the reboot stylesheet?

```css
iframe {
  color-scheme: light dark;
}
```

**Comment by dougwaldron** [comment]
> I think setting the following should suffice, and be relatively bulletproof for us without causing more downstream issues, being placed inside the reboot stylesheet?
> 
> iframe {
>   color-scheme: light dark;
> }

@mdo Thanks for reopening this! And good find with that article.

I put together a new CodePen that tests multiple scenarios here: https://codepen.io/dougwaldron/pen/KwaXoxY (Don't forget to test the page with your OS set to both light and dark mode and reload the page after changing!)


In the CodePen, I set the `color-scheme` of the parent element to light, dark, or not set, and did the same for the content inside the `iframe` (modeling what would come from a third-party component). Then I set the `color-scheme` of the `iframe` element itself to various values to model what could be set by Bootstrap. These include:

* no `color-scheme` set (current behavior)
* `normal`
* `light`
* `dark`
* `light dark`
* `normal light dark`
* `unset`

Unfortunately, your proposed fix (`light dark`) doesn't seem to work in all situations, particularly when the OS is set to dark mode, but I would appreciate someone double checking my work.

**Comment by mdo** [comment]
It looks correct here for me? https://v6-dev--twbs-bootstrap.netlify.app/docs/6.0/content/reboot/

Or is there another combination this doesn't account for?

**Comment by habibfyssal-bit** [comment]
I'll take a look at the Netlify link and compare it against the scenarios
in the CodePen to see if there are any remaining edge cases.

**Comment by dougwaldron** [comment]
> It looks correct here for me? https://v6-dev--twbs-bootstrap.netlify.app/docs/6.0/content/reboot/
> 
> Or is there another combination this doesn't account for?

In your examples, the content in each `iframe` sets its own color scheme to `light dark` using `<meta name='color-scheme' content='light dark'>`. 

But if your `iframe` is embedding a third-party component (e.g., Paypal, TrustPilot), I'd guess that more often than not, that component won't have a color scheme set at all. When the OS or browser is set to dark mode, and the `iframe` element has `color-scheme: light dark`, then the content will have an opaque white background (just as it does in Bootstrap v5):

<img width="688" height="174" alt="Image" src="https://github.com/user-attachments/assets/5121be61-d3a1-4aad-ae2b-564eb7076f36" />

I'd suggest using `iframe { color-scheme: normal; }` instead. In all cases where no color scheme is set by the embedded content, the `iframe` renders with a transparent background. 

The one consideration with using `color-scheme: normal` is that an `iframe` with dark-mode content will render with an opaque black background (even if the page/parent element is also in dark mode). But I'd argue that if the content is setting it's own color scheme to dark mode, then it should be setting its own background color as well, so the `iframe` background won't show anyway.

<img width="675" height="177" alt="Image" src="https://github.com/user-attachments/assets/2658936a-46db-4092-89a1-c915aecbe664" />

**Comment by salahmokatil-dotcom** [comment]
Thanks for the detailed breakdown and for testing those scenarios in the
CodePen.

Your point about third-party components (like PayPal or TrustPilot) not
having a color scheme set is a great catch. Given that color-scheme: light
dark still results in an opaque white background when the OS is in dark
mode, switching to iframe { color-scheme: normal; } seems like a more
robust approach to ensure transparency where it's most needed.

I agree that if a component explicitly sets its color scheme to dark, it
should ideally be responsible for its own background color anyway.

**Comment by mdo** [comment]
Coming back to this, it seems there's no one-size-fits-all solution that fixes this everywhere for all browsers?

I built a standalone repro with two embeds side by side. The first declares `<meta name="color-scheme" content="light dark">`, like a YouTube or CodePen embed. The second declares nothing, like most payment buttons and comment widgets. Then I ran every candidate value for `iframe { color-scheme: … }` with the OS in dark mode.

**WebKit**

**Chromium**

| `color-scheme` on `iframe` | Adaptive embed | Light-only embed |
| --- | --- | --- |
| `light dark` (v6 default) | transparent | white box |
| `normal` | transparent, but **forced light in WebKit** | transparent |
| `light` | forced light | transparent |
| `only light` | forced light | transparent |

`normal` is the value suggested in this thread. It does remove the white box. But WebKit propagates it into the embedded document, so every adaptive embed loses its dark mode. That trades a white box behind one widget for a broken theme on all the others, and only in Safari.

The white box is also not something we can style away. It is a browser safeguard. When the embedded document paints light content inside a dark page, the browser puts an opaque `canvas` background behind it so the content stays readable. `background: transparent` on the `iframe` does not defeat it.

So v6 keeps `color-scheme: light dark` in Reboot and adds a `.color-scheme-light` utility to opt out one embed at a time:

```html
<iframe class="color-scheme-light" src="..."></iframe>
```

This is the case where you already know the widget has no dark mode. Applying it globally would break the adaptive embeds instead. The Reboot and utility docs cover the mechanism and the WebKit caveat.

**Comment by mdo** [comment]
Webkit and Chromium:

<img width="1720" height="840" alt="Image" src="https://github.com/user-attachments/assets/87cb0674-45eb-4361-ba34-bf07f09acd1b" />

<img width="1720" height="840" alt="Image" src="https://github.com/user-attachments/assets/f2706b37-54cc-4775-9e22-2ae9c1939456" />

**Comment by androidacy-user** [comment]
Worth noting normal isn't a valid option iirc. Just makes the browser treat it as a custom and undefined color scheme. Technically works, but also technically incorrect

**Comment by dougwaldron** [comment]
> Coming back to this, it seems there's no one-size-fits-all solution that fixes this everywhere for all browsers?

That seems to be the case! I didn't test in WebKit, and I didn't expect it to work differently.

> So v6 keeps `color-scheme: light dark` in Reboot and adds a `.color-scheme-light` utility to opt out one embed at a time:
> 
> <iframe class="color-scheme-light" src="..."></iframe>
> This is the case where you already know the widget has no dark mode. Applying it globally would break the adaptive embeds instead. The Reboot and utility docs cover the mechanism and the WebKit caveat.

This sounds reasonable to me. Bootstrap page authors (us) will have to check whether any embedded `iframe` has a dark mode, but having an override for those situations will be welcome.

**Comment by wafir645fcqs65** [comment]
Thanks for confirming! That sounds like a sensible approach. Having a
dedicated utility for known light-only embeds while keeping adaptive
behavior by default seems like a good balance.

**Comment by mdo** [comment]
Fixed in v6 by #42605 and #42807, both now on `v6-dev`.

Reboot pins every `iframe` to `color-scheme: light dark`, so an embed stays transparent when you force a theme with `data-bs-theme`. For a third-party embed that declares no dark mode, the browser still paints an opaque canvas to keep its light content readable. Add the new `.color-scheme-light` utility to that one iframe.

There is no single value that fixes every case in every browser. `normal`, `light`, and `only light` are not options for the Reboot rule, because Safari and Firefox pass them into the embedded document and remove its dark mode. That is why the opt-out is per-element.

Thanks @dougwaldron for the repros and the testing across browsers, and everyone else who worked through the cases here.

Not backporting to v5, so I removed that label. v5 keeps `color-scheme` on the theme selectors only.
